File: src/main/java/org/uqbar/commons/model/ObservableUtils.java
package org.uqbar.commons.model;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

import org.uqbar.commons.utils.Observable;
import org.uqbar.commons.utils.ReflectionUtils;

/**
 * Utility class for {@link Observable} object functions, for example to force
 * triggering a property change method.
 *
 * @author jfernandes
 * @author flbulgarelli
 */
public class ObservableUtils {

	/**
	 * Set ups one or more dependencies for the given <code>property</code> in
	 * the given <code>model</code>. That is, the property <code>property</code>
	 * will fire a <code>propertyChanged</code> event whenever any of the
	 * <code>dependencies</code> get updated.
	 * 
	 * @since 3.4
	 * @param model
	 *            the model with the given properties
	 * @param property
	 *            the property to which dependencies will be added
	 * @param dependencies
	 *            the the dependencies to be added to the given property
	 */
	public static void dependencyOf(final Object model, final String property, String... dependencies) {
		addPropertyListener(model, new PropertyChangeListener() {
			public void propertyChange(PropertyChangeEvent evt) {
				firePropertyChanged(model, property);
			}
		}, dependencies);
	}

	public static void firePropertyChanged(Object model, String propertyName, Object newValue) {
		Method firePropertyChangedMethod = ReflectionUtils.findMethod(model.getClass(), "firePropertyChange", true, new Class[] { String.class, Object.class, Object.class });
		if(firePropertyChangedMethod != null){
			ReflectionUtils.invoke(model, firePropertyChangedMethod, propertyName, null, newValue);
		}
	}

	public static void firePropertyChanged(Object model, String property) {
		firePropertyChanged(model, property, ReflectionUtils.invokeGetter(model, property));
	}

	public static void addPropertyListener(Object model, PropertyChangeListener listener, String...properties) {
		addPropertyListener(model, listener, Arrays.asList(properties));
	}

	public static void addPropertyListener(Object model, PropertyChangeListener listener, List<String> properties) {
		for (String property: properties) {
			ReflectionUtils.invokeMethod(model, "addPropertyChangeListener", new Object[]{property, listener});
		}
	}

}

File: src/main/scala/org/uqbar/commons/utils/ObservableUtils.scala
package org.uqbar.commons.utils

import org.uqbar.commons.model.ObservableUtils
import java.beans.PropertyChangeListener
import java.beans.PropertyChangeEvent
import scala.collection.mutable.Buffer
import scala.collection.JavaConversions._
import scala.collection.JavaConverters._

case class When(val obj:Any){
  def change(property:String) = Change(property, this)
}
case class Change(property:String, val when:When){
  val properties = Buffer[String](property)
  def and(property:String) = {
    properties.append(property)
    this
  }
  def fireChange(property:String) = FireChange(property, this)
}
case class FireChange(val property:String, val change:Change){
  def of(obj:Any) {
    ObservableUtils.addPropertyListener(change.when.obj, new PropertyChangeListener(){
    	def propertyChange(evt:PropertyChangeEvent){
    	   ObservableUtils.firePropertyChanged(obj, property, ReflectionUtils.readField(obj, property)); 
    	}
    }, change.properties)
  }
}
